ksmt-cvc5/src/main/kotlin/org.ksmt.solver.cvc5/KCvc5ExprInternalizer.kt

```diff
@@ -0,0 +1,1033 @@
+        }
+    }
+
+    override fun <T : KBvSort> transform(expr: KBvArithShiftRightExpr<T>) = with(expr) {
+        transform(arg, shift) { arg: Term, shift: Term ->
+            nsolver.mkTerm(Kind.BITVECTOR_ASHR, arg, shift)
+        }
+    }
+
+    /*
+     * we can internalize rotate expr as concat expr due to simplification,
+     * otherwise we can't process rotate expr on expression
+     */
+    override fun <T : KBvSort> transform(expr: KBvRotateLeftExpr<T>) = with(expr) {
+        val simplifiedExpr = ctx.simplifyBvRotateLeftExpr(arg, rotation)
+
+        if (simplifiedExpr is KBvRotateLeftExpr<*>)
+            throw KSolverUnsupportedFeatureException("Rotate expr with expression argument is not supported by cvc5")
+
+        apply(simplifiedExpr)
+    }
+
+    /*
+     * @see transform(expr: KBvRotateLeftExpr<T>)
+     */
+    override fun <T : KBvSort> transform(expr: KBvRotateRightExpr<T>) = with(expr) {
+        val simplifiedExpr = ctx.simplifyBvRotateRightExpr(arg, rotation)
+
+        if (simplifiedExpr is KBvRotateRightExpr<*>)
+            throw KSolverUnsupportedFeatureException("Rotate expr with expression argument is not supported by cvc5")
+
+        apply(simplifiedExpr)
+    }
+
+    override fun <T : KBvSort> transform(expr: KBvRotateLeftIndexedExpr<T>) = with(expr) {
+        transform(value) { value: Term ->
+            val rotationOp = nsolver.mkOp(Kind.BITVECTOR_ROTATE_LEFT, rotationNumber)
+            nsolver.mkTerm(rotationOp, value)
+        }
+    }
+
+
+    override fun <T : KBvSort> transform(expr: KBvRotateRightIndexedExpr<T>) = with(expr) {
+        transform(value) { value: Term ->
+            val rotationOp = nsolver.mkOp(Kind.BITVECTOR_ROTATE_RIGHT, rotationNumber)
+            nsolver.mkTerm(rotationOp, value)
+        }
+    }
+
+    // custom implementation
+    @Suppress("MagicNumber")
+    override fun transform(expr: KBv2IntExpr) = with(expr) {
+        transform(value) { value: Term ->
+            // by default, it is unsigned in cvc5
+            val intTerm = nsolver.mkTerm(Kind.BITVECTOR_TO_NAT, value)
+
+            if (isSigned) {
+                val size = this.value.sort.sizeBits.toInt()
+                val modulo = powerOfTwo(size.toUInt())
+                val maxInt = (powerOfTwo((size - 1).toUInt())) - BigInteger.ONE
+
+                val moduloTerm = nsolver.mkInteger(modulo.toString(10))
+                val maxIntTerm = nsolver.mkInteger(maxInt.toString(10))
+
+                val gtTerm = nsolver.mkTerm(Kind.GT, intTerm, maxIntTerm)
+                val subTerm = nsolver.mkTerm(Kind.SUB, intTerm, moduloTerm)
+
+                nsolver.mkTerm(Kind.ITE, gtTerm, subTerm, intTerm)
+            } else intTerm
+        }
+    }
+
+    override fun <T : KBvSort> transform(expr: KBvAddNoOverflowExpr<T>) = with(expr) {
+        val simplifiedExpr = ctx.rewriteBvAddNoOverflowExpr(expr.arg0, expr.arg1, isSigned)
+        apply(simplifiedExpr)
```

Comment:
`apply` doesn't works correctly in this scenario

Reply:
Should the way below be used here instead of `apply` / `accept`? 

```
transform {
    expr.internalizeExpr() 
}
```

Reply:
No, you need smth like:
```kotlin
retun simplifiedExpr.accept(this)
```

Reply:
Ok, thank you